Adopt mypy (#6821)

hannes-ucsc opened the pull request:
<!--
This is the PR template for regular PRs against `develop`. Edit the URL in your
browser's location bar, appending either `&template=anvilprod-promotion.md`,
`&template=prod-promotion.md`, `&template=anvilprod-hotfix.md`, `&template=prod-
hotfix.md`, `&template=backport.md` or `&template=upgrade.md` to switch the
template.
-->

Connected issues: #6821


## Checklist


### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] On ZenHub, PR is connected to all issues it (partially) resolves
- [x] PR description links to connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] PR title references all connected issues
- [x] For each connected issue, there is at least one commit whose title references that issue

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title


### Author (partiality)

- [x] Added `p` tag to titles of partial commits
- [x] This PR is labeled `partial` <sub>or completely resolves all connected issues</sub>
- [x] This PR partially resolves each of the connected issues <sub>or does not have the `partial` label</sub>


### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or is not chained to another PR</sub>
- [x] The blocking PR is labeled `base` <sub>or this PR is not chained to another PR</sub>
- [x] This PR is labeled `chained` <sub>or is not chained to another PR</sub>


### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or the changes introduced by this PR will not require reindexing of any deployment</sub>
- [x] This PR is labeled `reindex:dev` <sub>or the changes introduced by it will not require reindexing of `dev`</sub>
- [x] This PR is labeled `reindex:anvildev` <sub>or the changes introduced by it will not require reindexing of `anvildev`</sub>
- [x] This PR is labeled `reindex:anvilprod` <sub>or the changes introduced by it will not require reindexing of `anvilprod`</sub>
- [x] This PR is labeled `reindex:prod` <sub>or the changes introduced by it will not require reindexing of `prod`</sub>
- [x] This PR is labeled `reindex:partial` and its description documents the specific reindexing procedure for `dev`, `anvildev`, `anvilprod` and `prod` <sub>or requires a full reindex or carries none of the labels `reindex:dev`, `reindex:anvildev`, `reindex:anvilprod` and `reindex:prod`</sub>
- [x] This PR and its connected issues are labeled `API` <sub>or this PR does not modify a REST API</sub>
- [x] Added `a` (`A`) tag to commit title for backwards (in)compatible changes <sub>or this PR does not modify a REST API</sub>
- [x] Updated REST API version number in `app.py` <sub>or this PR does not modify a REST API</sub>


### Author (upgrading deployments)

- [x] Ran `make docker_images.json` and committed the resulting changes <sub>or this PR does not modify `azul_docker_images`, or any other variables referenced in the definition of that variable</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading deployments</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading deployments</sub>
- [x] This PR is labeled `upgrade` <sub>or does not require upgrading deployments</sub>
- [x] This PR is labeled `deploy:shared` <sub>or does not modify `docker_images.json`, and does not require deploying the `shared` component for any other reason</sub>
- [x] This PR is labeled `deploy:gitlab` <sub>or does not require deploying the `gitlab` component</sub>
- [x] This PR is labeled `deploy:runner` <sub>or does not require deploying the `runner` image</sub>


### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the none of the stable branches (`anvilprod` and `prod`) have temporary hotfixes for any of the issues connected to this PR</sub>


### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not modify `requirements*.txt`, `common.mk`, `Makefile` and `Dockerfile`</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not modify `requirements*.txt`</sub>
- [x] This PR is labeled `reqs` <sub>or does not modify `requirements*.txt`</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not modify functionality that could affect the IT outcome</sub>


### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] PR is awaiting requested review from system administrator
- [x] PR is assigned to only the system administrator


### System administrator (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] A comment to this PR details the completed security design review
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved connected issues to *Approved* column
- [x] PR is assigned to only the operator


### Operator (before pushing merge the commit)

- [x] Checked `reindex:…` labels and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Ran `_select dev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select dev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Ran `_select anvildev.shared && CI_COMMIT_REF_NAME=develop make -C terraform/shared apply_keep_unused` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.gitlab && CI_COMMIT_REF_NAME=develop make -C terraform/gitlab apply` <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Checked the items in the next section <sub>or this PR is labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the system administrator <sub>or this PR is not labeled `deploy:gitlab`</sub>


### System administrator

- [x] Background migrations for [`dev.gitlab`](https://gitlab.dev.singlecell.gi.ucsc.edu/admin/background_migrations) are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] Background migrations for [`anvildev.gitlab`](https://gitlab.anvil.gi.ucsc.edu/admin/background_migrations) are complete <sub>or this PR is not labeled `deploy:gitlab`</sub>
- [x] PR is assigned to only the operator


### Operator (before pushing merge the commit)

- [x] Ran `_select dev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Ran `_select anvildev.gitlab && make -C terraform/gitlab/runner` <sub>or this PR is not labeled `deploy:runner`</sub>
- [x] Added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `dev`</sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices in `anvildev`</sub>
- [x] Started reindex in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR is not labeled `reindex:dev`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR is not labeled `reindex:anvildev`</sub>
- [x] The title of the merge commit starts with the title of this PR
- [x] Added PR # reference to merge commit title
- [x] Collected commit title tags in merge commit title <sub>but only included `p` if the PR is also labeled `partial`</sub>
- [x] Moved connected issues to *Merged lower* column in ZenHub
- [x] Moved blocked issues to *Triage* <sub>or no issues are blocked on the connected issues</sub>
- [x] Pushed merge commit to GitHub


### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>


### Operator (after pushing the merge commit)

- [x] Pushed merge commit to GitLab `dev`
- [x] Pushed merge commit to GitLab `anvildev`
- [x] Build passes on GitLab `dev`
- [x] Reviewed build logs for anomalies on GitLab `dev`
- [x] Build passes on GitLab `anvildev`
- [x] Reviewed build logs for anomalies on GitLab `anvildev`
- [x] Ran `_select dev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Ran `_select anvildev.shared && make -C terraform/shared apply` <sub>or this PR is not labeled `deploy:shared`</sub>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`


### Operator (reindex)

- [x] Deindexed all unreferenced catalogs in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed all unreferenced catalogs in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Deindexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Deindexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Indexed specific sources in `dev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:dev`</sub>
- [x] Indexed specific sources in `anvildev` <sub>or this PR is neither labeled `reindex:partial` nor `reindex:anvildev`</sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Checked for, triaged and possibly requeued messages in both fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Emptied fail queues in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Emptied fail queues in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Restarted the Data Browser pipeline for the [ucsc/hca/dev branch](https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/data-browser/-/pipelines/new?ref=ucsc%2Fhca%2Fdev) on GitLab in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Restarted the Data Browser pipeline for the [ucsc/lungmap/dev branch](https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/data-browser/-/pipelines/new?ref=ucsc%2Flungmap%2Fdev) on GitLab in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Restarted `deploy_browser` job in the GitLab pipeline for this PR in `dev` <sub>or this PR does not require reindexing `dev`</sub>
- [x] Restarted the Data Browser pipeline for the [ucsc/anvil/anvildev branch](https://gitlab.anvil.gi.ucsc.edu/ucsc/data-browser/-/pipelines/new?ref=ucsc%2Fanvil%2Fanvildev) on GitLab in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>
- [x] Restarted `deploy_browser` job in the GitLab pipeline for this PR in `anvildev` <sub>or this PR does not require reindexing `anvildev`</sub>


### Operator

- [x] Propagated the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `API`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels to the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] Propagated any specific instructions related to the `deploy:shared`, `deploy:gitlab`, `deploy:runner`, `API`, `reindex:partial`, `reindex:anvilprod` and `reindex:prod` labels, from the description of this PR to that of the next promotion PRs <sub>or this PR carries none of these labels</sub>
- [x] PR is assigned to no one


## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

hannes-ucsc committed:
Convert require()/reject() to assert

hannes-ucsc committed:
Replace typing.Optional and typing.Union

hannes-ucsc committed:
Replace `typing.Type` with `type`

hannes-ucsc committed:
Extract field types into separate module

hannes-ucsc committed:
[p 1/2] Cover azul.indexer.document with mypy (#6821)

Low hanging fruit. Incomplete (residual type errors).

hannes-ucsc committed:
[p 2/2] Cover azul.indexer.document with mypy (#6821)

Fix residual type errors.

hannes-ucsc committed:
[p 1/2] Cover azul.indexer.field with mypy (#6821)

Replace Generic, TypeVar and legacy type aliases

hannes-ucsc committed:
[p 2/2] Cover azul.indexer.field with mypy (#6821)

Significant refactoring to fix residual type errors.

hannes-ucsc committed:
[1/5] Refactor get_generic_type_params

Adopt PEP-695

hannes-ucsc committed:
[2/5] Refactor get_generic_type_params

Rename to derived_type_params

hannes-ucsc committed:
[4/5] Refactor get_generic_type_params

Make `root` argument to derived_type_params keyword-only

hannes-ucsc commented:
### Security design review

- [x] Security design review completed; this PR does **not** …
  - [x] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with Terra 
  - [x] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [x] … affect the permissions of external users like access to
    - TDR snapshots
  - [x] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [x] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [x] … affect monitoring of the system
  - [x] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [x] … add an interface that exposes sensitive or confidential data at the security boundary
  - [x] … affect the encryption of data at rest
  - [x] … require persistence of sensitive or confidential data that might require encryption at rest
  - [x] … require unencrypted transmission of data within the security boundary
  - [x] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [x] Documentation on any unchecked boxes is provided in comments below

achave11-ucsc commented:
[IT failed](https://gitlab.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/57997) in `anvilbox`. IT index notifications failed to index, they seemed to all have failed with the following error:

**CloudWatch Logs Insights**    
region: us-east-1    
log-group-names: /aws/lambda/azul-indexer-anvilbox, /aws/lambda/azul-indexer-anvilbox-contribute    
start-time: -3600s    
end-time: 0s    
query-string:
  ```
  fields @timestamp, @message
| filter @message like /ERROR|timed out/
| sort @timestamp asc
| limit 3
  ```
---
| @timestamp | @message |
| --- | --- |
| 2025-02-12 20:18:00.694 | [ERROR] TypeError: ('Root is not an ancestor', <class 'azul.indexer.SourcedBundleFQID'>, <class 'azul.plugins.repository.tdr_anvil.TDRAnvilBundleFQID'>) Traceback (most recent call last):   File "/var/task/azul/chalice.py", line 190, in patched_event_source_handler     return old_handler(self_, event, context)   File "/var/task/chalice/app.py", line 1761, in __call__     return self.handler(event_obj)   File "/var/task/app.py", line 193, in contribute     app.index_controller.contribute(event)   File "/var/task/azul/indexer/index_controller.py", line 173, in contribute     contributions, replicas = self.transform(catalog, notification, delete)   File "/var/task/azul/indexer/index_controller.py", line 225, in transform     bundle = service.fetch_bundle(catalog, bundle_fqid)   File "/var/task/azul/indexer/index_service.py", line 202, in fetch_bundle     bundle_fqid = plugin.bundle_fqid_from_json(bundle_fqid)   File "/var/task/azul/plugins/__init__.py", line 639, in bundle_fqid_from_json     return self._bundle_fqid_cls.from_json(fqid)   File "/var/task/azul/indexer/__init__.py", line 601, in from_json     source = cls.source_ref_cls().from_json(json['source'])   File "/var/task/azul/indexer/__init__.py", line 594, in source_ref_cls     ref_cls, = derived_type_params(cls, root=SourcedBundleFQID)   File "/var/task/azul/types.py", line 406, in derived_type_params     raise TypeError('Root is not an ancestor', root, cls) |
| 2025-02-12 20:18:02.909 | [ERROR] TypeError: ('Root is not an ancestor', <class 'azul.indexer.SourcedBundleFQID'>, <class 'azul.plugins.repository.tdr_anvil.TDRAnvilBundleFQID'>) Traceback (most recent call last):   File "/var/task/azul/chalice.py", line 190, in patched_event_source_handler     return old_handler(self_, event, context)   File "/var/task/chalice/app.py", line 1761, in __call__     return self.handler(event_obj)   File "/var/task/app.py", line 193, in contribute     app.index_controller.contribute(event)   File "/var/task/azul/indexer/index_controller.py", line 173, in contribute     contributions, replicas = self.transform(catalog, notification, delete)   File "/var/task/azul/indexer/index_controller.py", line 225, in transform     bundle = service.fetch_bundle(catalog, bundle_fqid)   File "/var/task/azul/indexer/index_service.py", line 202, in fetch_bundle     bundle_fqid = plugin.bundle_fqid_from_json(bundle_fqid)   File "/var/task/azul/plugins/__init__.py", line 639, in bundle_fqid_from_json     return self._bundle_fqid_cls.from_json(fqid)   File "/var/task/azul/indexer/__init__.py", line 601, in from_json     source = cls.source_ref_cls().from_json(json['source'])   File "/var/task/azul/indexer/__init__.py", line 594, in source_ref_cls     ref_cls, = derived_type_params(cls, root=SourcedBundleFQID)   File "/var/task/azul/types.py", line 406, in derived_type_params     raise TypeError('Root is not an ancestor', root, cls) |
| 2025-02-12 20:18:10.210 | [ERROR] TypeError: ('Root is not an ancestor', <class 'azul.indexer.SourcedBundleFQID'>, <class 'azul.plugins.repository.tdr_anvil.TDRAnvilBundleFQID'>) Traceback (most recent call last):   File "/var/task/azul/chalice.py", line 190, in patched_event_source_handler     return old_handler(self_, event, context)   File "/var/task/chalice/app.py", line 1761, in __call__     return self.handler(event_obj)   File "/var/task/app.py", line 193, in contribute     app.index_controller.contribute(event)   File "/var/task/azul/indexer/index_controller.py", line 173, in contribute     contributions, replicas = self.transform(catalog, notification, delete)   File "/var/task/azul/indexer/index_controller.py", line 225, in transform     bundle = service.fetch_bundle(catalog, bundle_fqid)   File "/var/task/azul/indexer/index_service.py", line 202, in fetch_bundle     bundle_fqid = plugin.bundle_fqid_from_json(bundle_fqid)   File "/var/task/azul/plugins/__init__.py", line 639, in bundle_fqid_from_json     return self._bundle_fqid_cls.from_json(fqid)   File "/var/task/azul/indexer/__init__.py", line 601, in from_json     source = cls.source_ref_cls().from_json(json['source'])   File "/var/task/azul/indexer/__init__.py", line 594, in source_ref_cls     ref_cls, = derived_type_params(cls, root=SourcedBundleFQID)   File "/var/task/azul/types.py", line 406, in derived_type_params     raise TypeError('Root is not an ancestor', root, cls) |
---

hannes-ucsc commented:
I tested this on a HCA-based personal deployment. I need to test this on an AnVIL-based one as well.

hannes-ucsc committed:
[3/5] Refactor get_generic_type_params

Remove optional runtime type checking

A function should do one thing, and one thing well. More importantly, having the checks outside the function benefits the type checker since the narrowed type information isn't discarded anymore by having a wider return type.

hannes-ucsc committed:
[5/5] Refactor get_generic_type_params

Fix handling of non-generic intermediate ancestors

hannes-ucsc committed:
[p] Cover azul.uuids, .indexer & .plugins with mypy (#6821)

hannes-ucsc commented:
Tested in both HCA- and AnVIL-based personal deployments.
